### PR TITLE
Do not silently fail in case of class scanning exceptions

### LINKED ISSUE #97: Do not silently fail in SpaceScanner when classes could not be parsed
Currently issues with parsing class files are only issued on DEBUG level (https://github.com/eclipse/sisu.inject/blob/a9752361662edac045f16129a49e1a8b625aa604/org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SpaceScanner.java#L115). As this is a frequent cause of trouble (https://lists.apache.org/thread/6wxrhlkrqsl7wgr730xd50h55t7ff0vy) this should at least optionally lead to an exception or at least log entry with ERROR level.

### PR DESCRIPTION
**Pull request by kwin**
This closes #97

### CODE REVIEW

**Comment by kwin** [comment]
@cstamas Any idea how to make this configurable from Maven? Do we really want to extend the Plexus Container API for that?

File: org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SisuIndex.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public final void index( final ClassSpace _space )
     {
         try
         {
-            new SpaceScanner( _space ).accept( this );
+            new SpaceScanner( _space, true ).accept( this );
```
Comment (kwin) [inline org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SisuIndex.java:95]: I guess being always strict here should be fine

File: org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SpaceModule.java [hunk 2]
```diff
@@ -166,7 +166,7 @@ public SpaceVisitor visitor( final Binder binder )
 
     void scanForElements( final Binder binder )
     {
-        new SpaceScanner( space, finder ).accept( strategy.visitor( binder ) );
+        new SpaceScanner( space, finder, true ).accept( strategy.visitor( binder ) );
```
Comment (kwin) [inline org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SpaceModule.java]: this should probably expose a way on how to toggle the strict flag.

**Review by mcculls — CHANGES_REQUESTED** [review]
I think the default mode should remain non-strict, for reasons discussed in https://github.com/eclipse/sisu.inject/issues/97#issuecomment-1774217953

We can still change the mode to strict when generating the index via annotation processing or the maven plugin. That should solve the main build-time issue without affecting runtime deployments.

**Comment by kwin** [comment]
@mcculls Can you be more specific, where exactly you want to keep the default behaviour of being lenient?

1. https://github.com/eclipse/sisu.inject/blob/d2a7e64ad07a026ad488b042a37744e66ff9a75e/org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SisuIndex.java#L95
2. https://github.com/eclipse/sisu.inject/blob/d2a7e64ad07a026ad488b042a37744e66ff9a75e/org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/SpaceModule.java#L169
3. https://github.com/eclipse/sisu.inject/blob/d2a7e64ad07a026ad488b042a37744e66ff9a75e/org.eclipse.sisu.inject/src/main/java/org/eclipse/sisu/space/QualifierCache.java#L83

or anywhere else (which I missed in the list above)

**Comment by kwin** [comment]
@mcculls Can you comment on the individual cases listed above?

**Comment by mcculls** [comment]
`SisuIndex` can be strict by default because it's build-time - everything else should be non-strict by default, with an option to make them strict.

**Comment by kwin** [comment]
> everything else should be non-strict by default, with an option to make them strict.

@mcculls done now, can you check again?

**Comment by kwin** [comment]
If there are no further concerns @mcculls I am gonna merge end of this week.

**Pull request merged**